utils/training_loop.py

```diff
@@ -1,270 +1,104 @@
-#             steps_counter += 1
-#
-#             # Perform a step
-#             states, actions, next_states, collision_occurred, reached_target, reward = rl.step_agent_only()
-#
-#             # Add current step to trajectory (once for car1 action, and once for car2 (both have same reward))
-#             rl.current_trajectory.append((states, actions, next_states, reward))
-#
-#             # if config.TRAIN_OPTION == 'step':
-#             #     step_loss = rl.train_trajectory(train_only_last_step=True, episode_counter=episode_counter)
-#             #     logger.log_scaler("loss_per_step", steps_counter, step_loss)
-#
-#             # Update the sum of rewards
-#             episode_sum_of_rewards += reward
-#
-#             # Handle collision or target reached scenarios
-#             if collision_occurred or reached_target:
-#                 airsim.reset_cars_to_initial_positions()
-#
-#                 # Log the episode's sum of rewards
-#                 logger.log_scaler("episode_sum_of_rewards", episode_counter, episode_sum_of_rewards)
-#
-#                 # Update collision count and break the loop
-#                 if collision_occurred:
-#                     collision_counter += 1
-#                 break
-#
-#         # Update epsilon greedy after each trajectory
-#         rl.epsilon *= rl.epsilon_decay
-#         print(f"epsilon value: {rl.epsilon}")
-#
-#         if config.TRAIN_OPTION == 'trajectory' and not config.ONLY_INFERENCE:
-#             trajectory_loss = rl.train_trajectory(train_only_last_step=False, episode_counter=episode_counter)
-#             logger.log_scaler("loss_per_trajectory", episode_counter, trajectory_loss)
-#             if config.LOG_SAME_ACTION_SELECTED_IN_TRAJECTORY:
-#                 logger.log_same_action_selected_in_trajectory()
-#
-#         rl.trajectories.append(rl.current_trajectory)
-#
-#         # TODO: complete train_batch_of_trajectories
-#         # if config.TRAIN_OPTION == 'batch_of_trajectories':
-#         #     if episode_counter % batch_size:
-#         #         agent.train_batch_of_trajectories(rl.trajectories)
-#
-#     # Save the network weights after the experiment ended (name of weights file is set in config)
-#     nn_handler.save_network_weights(rl.network)
-#
-#     # reset cars to initial settings file position, before the next experiment
-#     airsim.reset_cars_to_initial_settings_file_positions()
-#
-#     print("---- Experiment Ended ----")
-#     print("The amount of collisions:")
-#     print(collision_counter)
-
+        print(f"Episode {episode_counter} finished with reward: {episode_sum_of_rewards}")
+        all_rewards.append(episode_sum_of_rewards)
+        steps_counter = 0
+    model.save(path + '/model')
+    print('Model saved')
+    print("Total collisions:", collision_counter)
+    loss_data = pd.read_csv(f"{path}/progress.csv")
+    losses = loss_data["train/value_loss"]
+    plt.figure(figsize=(12, 6))
```

Comment:
move this code to a function (maybe add plotting utils...

Reply:
fixed